### PR TITLE
update `fort.15` and driver generation

### PR DESCRIPTION
**Pull request by ghost**
update `fort15.py` and `driver.py`

### CODE REVIEW

File: adcircpy/forcing/base.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from abc import ABC, abstractmethod
+from os import PathLike
+
+
+class Forcing(ABC):
+    @abstractmethod
+    def write(self, directory: PathLike, overwrite: bool = False):
+        raise NotImplementedError
```
Comment (ghost) [inline adcircpy/forcing/base.py:8]: abstract forcing class

File: adcircpy/forcing/waves/base.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class WaveForcing(Forcing):
+    def __init__(self, nrs: int, interval_seconds: int):
+        self.NRS = nrs
+        self.RSTIMINC = interval_seconds
+
+    def write(self, directory: PathLike, overwrite: bool = False):
+        # TODO implement this
+        pass
```
Comment (ghost) [inline adcircpy/forcing/waves/base.py]: wave forcing class (incomplete)

File: adcircpy/forcing/waves/base.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from os import PathLike
+
+from adcircpy.forcing.base import Forcing
```
Comment (ghost) [inline adcircpy/forcing/waves/base.py:4]: should be relative import

File: adcircpy/forcing/winds/base.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class WindForcing(Forcing):
+    def __init__(self, nws: int, interval_seconds: int):
+        self.NWS = nws
+        self.WTIMINC = interval_seconds
+
+    def write(self, directory: PathLike, overwrite: bool = False):
+        # TODO implement this
+        pass
```
Comment (ghost) [inline adcircpy/forcing/winds/base.py]: abstract wind forcing class (incomplete)

File: adcircpy/forcing/winds/base.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from os import PathLike
+
+from adcircpy.forcing.base import Forcing
```
Comment (ghost) [inline adcircpy/forcing/winds/base.py:4]: should be relative import

File: adcircpy/fort15.py [hunk 6]
```diff
@@ -1,246 +1,283 @@
...
         # ----------------
-        f = f'{self.RUNDES}'.ljust(63) + ' ! RUNDES\n'
-        f += f'{self.RUNID}'.ljust(63) + ' ! RUNID\n'
-        f += f'{self.NFOVER:d}'.ljust(63) + ' ! NFOVER\n'
-        f += f'{self.NABOUT:d}'.ljust(63) + ' ! NABOUT\n'
-        f += f'{self.NSCREEN:d}'.ljust(63) + ' ! NSCREEN\n'
-        f += f'{self.IHOT:d}'.ljust(63) + ' ! IHOT\n'
-        f += f'{self.ICS:d}'.ljust(63) + ' ! ICS\n'
-        f += f'{self.IM:d}'.ljust(63) + ' ! IM\n'
+        f = []
```
Comment (ghost) [inline adcircpy/fort15.py:27]: compile a list of lines, instead of concatenating to a string

File: adcircpy/fort15.py [hunk 7]
```diff
@@ -1,246 +1,283 @@
...
+        f.extend([
+            f'{self.RUNDES:<63} ! RUNDES',
```
Comment (ghost) [inline adcircpy/fort15.py:29]: use f string formatting to justify content

File: adcircpy/fort15.py [hunk 8]
```diff
@@ -582,15 +645,50 @@ def NSCREEN(self):
...
-            return 0
-        else:
-            if self.wind_forcing is not None:
+            nws = 0
+        elif self.mesh is not None:
+            wind_forcing = self.mesh._surface_forcing['imetype']
+            if wind_forcing is not None:
                 # check for wave forcing here as well.
-                return self.wind_forcing.NWS
+                nws = int(wind_forcing.NWS % 100)
```
Comment (ghost) [inline adcircpy/fort15.py:660]: extract the tens + ones places from the NWS parameter

File: adcircpy/fort15.py [hunk 9]
```diff
@@ -582,15 +645,50 @@ def NSCREEN(self):
...
+
+        if self._runtype == 'coldstart':
+            nrs = 0
+        elif self.mesh is not None:
+            wind_forcing = self.mesh._surface_forcing['imetype']
+            wave_forcing = self.mesh._surface_forcing['iwrtype']
+            if wave_forcing is not None:
+                nrs = wave_forcing.NRS
+            elif wind_forcing is not None:
+                nrs = int(math.floor(wind_forcing.NWS / 100))
```
Comment (ghost) [inline adcircpy/fort15.py:685]: extract the hundredth's place from NWS as the NRS parameter

File: adcircpy/fort15.py [hunk 10]
```diff
@@ -883,17 +991,27 @@ def WTIMINC(self):
...
 
+    @property
+    def RSTIMINC(self):
+        if self.NRS in [1, 3, 4, 5]:
+            if self.wave_forcing is not None:
+                return self.wave_forcing.RSTIMINC
+            else:
+                return self.WTIMINC
+        else:
+            return 0
```
Comment (ghost) [inline adcircpy/fort15.py:1002]: method for getting radiative stress time interval

File: adcircpy/mesh/mesh.py [hunk 11]
```diff
@@ -174,6 +175,8 @@ def add_forcing(self, forcing, id=None):
                 self._boundary_forcing[forcing.btype].update({"obj": forcing})
         elif isinstance(forcing, WindForcing):
             self._surface_forcing.update({'imetype': forcing})
+        elif isinstance(forcing, WaveForcing):
+            self._surface_forcing.update({'iwrtype': forcing})
```
Comment (ghost) [inline adcircpy/mesh/mesh.py:179]: include waveforcing

File: tests/test_configuration.py [hunk 12]
```diff
@@ -69,6 +71,48 @@ def test_slurm_driver(self):
             with open(reference_directory / 'slurm.job') as reference_file:
                 assert generated_file.read() == reference_file.read()
 
+    def test_configuration(self):
```
Comment (ghost) [inline tests/test_configuration.py:74]: add test

File: tests/test_configuration.py [hunk 13]
```diff
@@ -69,6 +71,48 @@ def test_slurm_driver(self):
...
+        )
+
+        driver.write(output_directory, overwrite=True)
+
+        for reference_filename in reference_directory.iterdir():
+            generated_filename = output_directory / reference_filename.name
+            with open(generated_filename) as generated_file, \
+                open(reference_filename) as reference_file:
+                assert generated_file.readlines()[1:] == \
+                       reference_file.readlines()[1:]
```
Comment (ghost) [inline tests/test_configuration.py]: check reference files are the same after the first line

File: adcircpy/mesh/mesh.py [hunk 14]
```diff
@@ -23,12 +24,12 @@ class AdcircMesh(EuclideanMesh2D):
     """
 
     def __init__(
-            self,
-            nodes,
-            elements,
-            boundaries=None,
-            crs=None,
-            description=None,
+        self,
```
Comment (ghost) [inline adcircpy/mesh/mesh.py]: I thought we had decided to do double indentations on call signatures based on a PEP rule ?

File: adcircpy/forcing/waves/base.py (hunk 2 above)
Comment (jreniel) [inline adcircpy/forcing/waves/base.py]: `raise NotImplementedError` ?

File: adcircpy/forcing/waves/base.py (hunk 3 above)
Comment (jreniel) [reply adcircpy/forcing/waves/base.py:4]: I'm actually favoring absolutes now, let's discuss this.

File: adcircpy/mesh/mesh.py (hunk 14 above)
Comment (ghost) [reply adcircpy/mesh/mesh.py]: yes you're right, fixed

**Pull request merged**